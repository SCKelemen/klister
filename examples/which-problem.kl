#lang "prelude.kl"

(import (shift "prelude.kl" 1))
(import (shift "quasiquote.kl" 1))
(import "quasiquote.kl")
(import "define-syntax-rule.kl")


(define-macros
  ([m (lambda (stx)
        (>>= (which-problem)
          (lambda (prob)
            (case prob
              [(declaration)  (pure '(example (the (m) (m))))]
              [(type)         (pure 'Bool)]
              [(expression t) (pure 'true)]
              [(pattern)      (pure 'unit)]))))]))

(m)

(example (case unit
           ((m) true)))


(define-macro (mega-const e)
  (>>= (which-problem)
       (lambda (prob)
          (case prob
            [(expression t)
             (type-case t
               [(-> a b)
                (pure `(lambda (_)
                         (mega-const ,e)))]
               [(else x) (pure e)])]))))

(example (the (-> Bool Bool Bool Bool Unit) (mega-const unit)))

(define-syntax-rule (llet (x value) body)
  ((lambda (x) body) value))

(example (llet (x (mega-const unit)) (the (-> Bool Bool Bool Bool Unit) x)))


(datatype (Both A) (both A A))

(example (both (lambda (x) (lambda (y) (lambda (z) 'hello))) (mega-const 'world)))
(example (both (mega-const 'hello) (lambda (x) (lambda (y) (lambda (z) 'world)))))

      
