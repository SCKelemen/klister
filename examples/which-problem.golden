(true) : Bool
(true) : Bool
#<closure> : (Bool → (Bool → (Bool → (Bool → Unit))))
#<closure> : (Bool → (Bool → (Bool → (Bool → Unit))))
(both #<closure> #<closure>) : ∀α β γ. (Both (α → (β → (γ → Syntax))))
(both #<closure> #<closure>) : ∀α β γ. (Both (α → (β → (γ → Syntax))))
